In [33]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import statsmodels.api as sm
from sklearn.model_selection import train_test_split
import warnings
warnings.filterwarnings("ignore")

from ggplot import *
from scipy.stats import logistic
from sklearn.metrics import roc_curve
from sklearn.metrics import roc_auc_score
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix, mean_squared_error
from sklearn.model_selection import train_test_split
from sklearn import preprocessing
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import *
from sklearn.tree import DecisionTreeClassifier, DecisionTreeRegressor, export_graphviz
import graphviz
from sklearn.ensemble import RandomForestRegressor, RandomForestClassifier
from sklearn.svm import SVC, SVR

# Lecture note 6 Classification using Naive Bayes

## 2.3 Bayes' Theorem

In [34]:
(4/20)*0.2/0.05 

0.8000000000000002

In [35]:
(0.2*0.9)/(0.2*0.9+0.8*0.3)

0.4285714285714286

In [36]:
(0.2*4/20)/(0.2*4/20+0.8*1/80)

0.7999999999999999

$\Pr(W_1|Spam)\times\Pr(\sim W_2|Spam)\times\Pr(\sim W_3|Spam)\Pr(W_4|Spam)$

In [37]:
(4/20)*(10/20)*(20/20)*(12/20)
# likelihood

0.06

## 정리

- Gaussian Distribution = Normal Distribution

# Gaussian NB
## x 변수가 연속형 변수 일 때 사용한다.

In [38]:
file1="c:/temp/iris.csv"
data1=pd.read_csv(file1)
data1

Sepal.Length  Sepal.Width  Petal.Length  Petal.Width    Species
0             5.1          3.5           1.4          0.2     setosa
1             4.9          3.0           1.4          0.2     setosa
2             4.7          3.2           1.3          0.2     setosa
3             4.6          3.1           1.5          0.2     setosa
4             5.0          3.6           1.4          0.2     setosa
..            ...          ...           ...          ...        ...
145           6.7          3.0           5.2          2.3  virginica
146           6.3          2.5           5.0          1.9  virginica
147           6.5          3.0           5.2          2.0  virginica
148           6.2          3.4           5.4          2.3  virginica
149           5.9          3.0           5.1          1.8  virginica

[150 rows x 5 columns]

In [39]:
pd.value_counts(data1['Species'])

setosa        50
versicolor    50
virginica     50
Name: Species, dtype: int64

In [40]:
# 데이터를 보면 x 변수들은 연속형 변수인 것을 알 수 있다

구하려는 것

$ \Pr(species=setosa|x1=x1,x2=x2,x3=x3,x4=x4)$
$ \Pr(species=virginica|x1=x1,x2=x2,x3=x3,x4=x4)$
$ \Pr(species=versicolor|x1=x1,x2=x2,x3=x3,x4=x4)$

이렇게 각각 계산해서 가장 높은 걸로 예측된다.

위 세개를 계산 할 때 Naive Bayes를 사용한다.

각 x 들은 normal 따른다고 가정

In [41]:
# Normal NB (Gaussian NB)

# train test set 하지말고 전체 표본을 가지고 예측 해보자

# 먼저 x, y 변수 지정
X=data1[['Sepal.Length','Sepal.Width','Petal.Length','Petal.Width']]
y=data1['Species']

# Gaussian NB 실행
gnb=GaussianNB().fit(X,y)
# gnb가 한 것: 150개의 데이터에 대해서 각각이 setos일 확률, versicolor일 확률, virginica 일 확률을 계산한 것

# 가장 확률 높은 것을 예측 해라 (?)
y_pred=gnb.predict(X)
confusion_matrix(y,y_pred)

array([[50,  0,  0],
       [ 0, 47,  3],
       [ 0,  3, 47]], dtype=int64)

#### confusion matrix 해석

- 50 0 0 : setosa는 50개 다 맞춤
- 0 47 3 : versicolor는 47개 맞추고, 3개는 virginica로 예측
- 0 3 47 : virginca 47개 맞추고, 3개는 versicolor로 잘못 예측
- ccr = 정확히 맞춘 비율

In [42]:
# CCR <- 여기선 'NB 방법으로 예측한 분류 예측의 성과'
(144/150)

0.96

# Multinomial NB
## x변수가 count 형일 때 사용한다.
- count 형 : 정수형태로 x 값이 떨어지는 데이터일 때
    - 아이가 몇명이냐? 3명 (3.5명이런건 없음.)




# Bernoulli NB
## x변수가 0/1 (-> 이향형)로 나타날때 사용한다.

# 강의안 정리

- 조건부 확률
- Bayes' Theorem
- Posterior probability
- Prior probability